### ISSUE TITLE
function decorator deficiency

### ISSUE THREAD

**Issue by pavlis** [mspass-team/mspass#354]
I discovered what I consider a deficiency in our processing function decorator (`mspass_func_wrapper`).  It has a useful feature wherein the function call is wrapped in a try block with generic exception handlers for RuntimeError and MsPASSError exceptions.   I think there are two problems with the current implementation:
1.  The one that is biting me now is that when a function throws an exception the error handlers actually do catch, they simply log the error and return the original data without a change.   That is not the right behavior.   It produced mysterious behavior for me, for example, because I ran the "rotate_to_standard" function, which is a wrapper for a method of Seismogram, with some real data that had a singular transformation matrix (not sure why but probably a metadata issue).   The algorithm blundered on and caused mysterious downstream behavior that it took a bit to sort out.  
2. I think the catch (except) block needs a generic catch all clause at the end.  I don't remember the syntax but @wangyinz you have advocated using this before.   I think we need a closing block for the exception handlers that can do that to avoid downstream error I got - "Unhandled exception" that added to the mystery. 

Item 1 can, I think, be solved by just adding a boolean argument with a name like `kill_if_failure` or we could get fancy and define an level for the `ErrorSeverity` enum class.   There is some reason for that because sometimes an algorithm might issue a "complaint" which may or may need to be fatal.   "information" should definitely not kill the datum.  So, fancy may be necessary.

Item 2 should be easy to fix and should be done immediately as it would probably not effect anything.  Item 1 will take some testing and will require edits to any existing function using the function decorator.

**Comment by pavlis** [comment]
Minor addition.   The decorator for the rotate_to_standard function I'm using is, I think, catching an error thrown by rotate_to_standard as a RuntimeError and I don't understand why we get that behavior.   Here is the error section in rotate_to_standard I know is being executed because the print (cout) line of scaffolding is being posted to the output while this is running:
```
        int three(3);
        dgetrf(three,three,a,three,ipivot,info);
        if(info!=0)
        {
            std::cout << "dgetrf returned info="<<info<<std::endl;
            for(i=0; i<3; ++i) delete [] work[i];
            throw(MsPASSError(
                      string("rotate_to_standard:  LU factorization of transformation matrix failed")),
                  ErrorSeverity::Invalid);
        }
```
But here is the output I'm seeing:
```
Caught an unknown exception! ErrorSeverity.Invalid
dgetrf returned info=1
dgetrf returned info=1
Caught an unknown exception! ErrorSeverity.Invalid
dgetrf returned info=1
dgetrf returned info=1
Caught an unknown exception! ErrorSeverity.Invalid
dgetrf returned info=1
dgetrf returned info=1
Caught an unknown exception! ErrorSeverity.Invalid
dgetrf returned info=2
dgetrf returned info=2
```
where the message about "unknown exception" comes from this python code in the workflow driving this processing:
```
for doc in cursor:
    d = db.read_data(doc,collection="wf_Seismogram")
    d = rotate_to_standard(d)
    if d.elog.size()>0:
        for x in d.elog.get_error_log():
            print(x.message,x.badness)
```
The decorator is handling this to create that log entry here:
```
    except RuntimeError as err:
        if isinstance(data, (Seismogram, TimeSeries)):
            data.elog.log_error(alg_name, str(err), ErrorSeverity.Invalid)
        else:
            logging_helper.ensemble_error(data, alg_name, err, ErrorSeverity.Invalid)
        # some unexpected error happen, if inplace_return is true, we may want to return the original data
        if inplace_return:
            return data
    except MsPASSError as ex:
        if ex.severity == ErrorSeverity.Fatal:
            raise
        if isinstance(data, (Seismogram, TimeSeries)):
            data.elog.log_error(alg_name, ex.message, ex.severity)
        else:
            logging_helper.ensemble_error(data, alg_name, ex.message, ex.severity)
        # some unexpected error happen, if inplace_return is true, we may want to return the original data
        if inplace_return:
            return data
```
The mystery here is that the exception block is properly being entered because the dgetdf function is returning an error code.  However, the decorator is not catching a MsPASError but seems to be just getting a RuntimeError with the generic message "Caught an unknown exception! " .   I do not know how that happens.   The only hypothesis I can make is that I remember @wangyinz puzzling out the way to map a MsPASSError into an exception that could be handled by that name instead of the generic "RuntimeError".  The hypothesis I have is that maybe that magic is broken when run through the decorator?   That may be totally wrong but I  have not other explanation.   

Any ideas how I can sort this out?   I'm not sure even going step by step through the mixed python-C++ code with gdb will help.

Incidentally, this is more a blemish than a serious issue once we address the need to have a complete trap for all possible exception types in the decorator.  It does the right thing but only a vague message gets posted to the error log.

**Comment by wangyinz** [comment]
Are you running this inside the container or locally? What is compiler used? I know that the mapping of C++ exception to Python can break on MacOS, and I suspect that it is due to clang, but I never get the problem sorted out. I think at least for the code inside the container compiled using gcc does not have such an issue. I did a search through our code base and can see that the "Caught an unknown exception" string is from pybind11 at `pybind11-src/include/pybind11/detail/internals.h:240`. How it runs to that line will need some investigation through a debugger for sure.

For the two items, I think we didn't implement a kill in the decorator because we thought that it should be done by the function itself. In other words, the wrapped function should decide whether to kill the data when an error occurs. Otherwise, as you said, we will have to look at the error severity in the decorator function, but still the decorator is agnostic on how bad an error is. So, I think it will be cleaner if we can just handle the kill inside `rotate_to_standard` (and probably all other functions that have it missing right now). What do you think?

**Comment by pavlis** [comment]
I was running this with my ubuntu machine with a locally compiled copy.   Thus it is not a clang problem.   I'm not sure how to sort this out even if I resort to gdb, but that seems the only solution.  Definitely will be easier to solve with a test program that removes some of the other complex baggage in this workflow.   If i can't recreate it in smaller test script it says something else.  I'll start with that step and let you know what I find.

The problem I see with not having a kill option in the decorator is that when I run the (decorated) rotate_to_standard as a function there is no way to tell if the output has failed unless I scan the error log.   What you see above does that but it takes no action other than printing the contents.   If I passed a datum that already had an entry in elog before running rotate_to_standard it would be even harder as I'd have to take apart all the entries and search for a particular error message.  Not exactly something we should ask a user to do.   

What the user could choose is if they want to consider handling errors more carefully.   I recommend the specific example here that is now this:
```
@mspass_func_wrapper
def rotate_to_standard(
    data,
    *args,
    object_history=False,
    alg_name=None,
    alg_id=None,
    dryrun=False,
    inplace_return=True,
    function_return_key=None,
    **kwargs
):
```
be changed to this
```
@mspass_func_wrapper
def rotate_to_standard(
    data,
    *args,
    object_history=False,
    alg_name=None,
    alg_id=None,
    dryrun=False,
    inplace_return=True,
    function_return_key=None,
    kill_on_error=True,
    **kwargs
):
```
Other functions may want to use `kill_on_error=False` and take different action for different errors.   In the actual use where the error is posted when the decorated function throws an exception it will likely be rare for the handler to want to do anything else but kill the result.    Seems to me if some needs a more subtle error handling they would need to rethink how they are handling the errors and use a different mechanism than throwing an exception.  In all books I've read throw/raise is intended to mean the algorithm failed and the result is not reliable.  
```

**Comment by wangyinz** [comment]
Yeah, I think that is reasonable.

For the RuntimeError issue, that is definitely surprising to me. I guess I need to also look into the problem then. Could you please tell me the Ubuntu version, compiler version, and Python version that has the issue? Also, I may need a small reproducer (if the error only occurs in certain calls).

**Comment by wangyinz** [comment]
Well, it seems the versions match what we have in the containers. Maybe you can try and see if the same issue occurs inside the container. I hope it doesn't...

**Comment by wangyinz** [comment]
Got it fixed already. Turns out it is not in the implementation of error handling, and the error message is telling exactly the truth - there is an unknown exception thrown. You can see what I did in https://github.com/mspass-team/mspass/commit/1637110c1bfc1ea9dab300d56e001366b5edbf09, but it is basically a mismatch of parenthesis that caused an unexpected error.

**Comment by pavlis** [comment]
Wow, was that subtle.   The kind of thing a human eye can't see easily and you expect the compiler to catch.   I guess I can see why the compiler might plow through that mistake, but not completely.   I'm amazed you even caught the detail.  I'm terrible about spotting that kind of thing.

That solved, we do still need to fix the kill_on_exception.   Not sure I want to take that on as it will need some tedious work to add tests for all decorated functions in algorithms.

**Comment by wangyinz** [comment]
Yeah... I spent a lot of effort stepping through the relevant code in gdb before realizing what it is. 

I may just fix the decorators but if I don't have the time, let's discuss it in the meeting and see who may take it.

**Comment by wangyinz** [comment]
The decorator behavior now matches the requested failure semantics. `mspass_func_wrapper` catches `RuntimeError`, non-Fatal `MsPASSError`, and a final generic `Exception`; each path logs the error, kills the datum, and returns it, while Fatal `MsPASSError` is re-raised. The wrapper tests cover these error paths. Closing as completed.
